File: integrations/hermes/open_android_intelligence_gateway/credentials.py
"""Account password credentials for the Hermes Gateway.

The password is the only secret a human enters, and contract section 5.2 keeps
it out of Android entirely: it is presented once over verified TLS and verified
by the Gateway. This module stores a stdlib-only scrypt digest so the plugin
keeps its empty dependency list, and it never stores or returns the password
itself.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import secrets

ALGORITHM = "scrypt"
# Interactive-login cost: ~16 MiB of memory, one pass, well under a second.
COST_N = 2 ** 14
BLOCK_R = 8
PARALLEL_P = 1
SALT_BYTES = 16
KEY_BYTES = 32
# scrypt hashes at most 1024 bytes of input (RFC 7914); longer input is a
# caller bug we refuse rather than silently truncate.
MAX_PASSWORD_BYTES = 1024


class CredentialFormatError(ValueError):
    """The stored credential is not in the one format this module writes."""


def _encode(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _decode(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def _password_bytes(password: str) -> bytes:
    if not isinstance(password, str) or not password:
        raise ValueError("password must be a non-empty string")
    encoded = password.encode("utf-8")
    if len(encoded) > MAX_PASSWORD_BYTES:
        raise ValueError("password is too long")
    return encoded


def _derive(password: str, salt: bytes, cost: int, block: int, parallel: int) -> bytes:
    return hashlib.scrypt(
        _password_bytes(password),
        salt=salt,
        n=cost,
        r=block,
        p=parallel,
        dklen=KEY_BYTES,
        maxmem=132 * cost * block,
    )


def hash_password(
    password: str, *, cost: int = COST_N, block: int = BLOCK_R, parallel: int = PARALLEL_P,
) -> str:
    """Returns a self-describing digest: ``scrypt$n$r$p$salt$key``."""
    salt = secrets.token_bytes(SALT_BYTES)
    key = _derive(password, salt, cost, block, parallel)
    return "$".join((ALGORITHM, str(cost), str(block), str(parallel), _encode(salt), _encode(key)))


def verify_password(password: str, encoded: str) -> bool:
    """Constant-time verification of a stored digest.

    Any malformed stored value, any unknown algorithm and any non-string input
    is a failed verification rather than an exception: the caller is an
    authentication seam that must fail closed.
    """
    if not isinstance(encoded, str) or not encoded:
        return False
    parts = encoded.split("$")
    if len(parts) != 6 or parts[0] != ALGORITHM:
        return False
    try:
        cost = int(parts[1])
        block = int(parts[2])
        parallel = int(parts[3])
        salt = _decode(parts[4])
        expected = _decode(parts[5])
    except (ValueError, TypeError):
        return False
    if cost < 2 or block < 1 or parallel < 1 or not salt or len(expected) != KEY_BYTES:
        return False
    try:
        candidate = _derive(password, salt, cost, block, parallel)
    except (ValueError, TypeError, MemoryError):
        return False
    return hmac.compare_digest(candidate, expected)

File: integrations/openclaw/src/core/credentials.ts
import { randomBytes, scryptSync, timingSafeEqual } from "node:crypto";

/**
 * Account password digests for the OpenClaw Gateway.
 *
 * Contract §5.2 keeps the password out of Android entirely: it is presented once
 * over verified TLS and verified here. Only a digest is stored, so the Gateway
 * can never read a password back, and an account that was never given one cannot
 * be logged into at all.
 */
export const ALGORITHM = "scrypt";
// Interactive-login cost: ~16 MiB of memory, one pass, well under a second.
export const COST_N = 2 ** 14;
export const BLOCK_R = 8;
export const PARALLEL_P = 1;
const SALT_BYTES = 16;
const KEY_BYTES = 32;

// scrypt hashes at most 1024 bytes of input (RFC 7914).
const MAX_PASSWORD_BYTES = 1024;

const encode = (value: Uint8Array): string => Buffer.from(value).toString("base64url");
const decode = (value: string): Buffer => Buffer.from(value, "base64url");

const passwordBytes = (password: string): Buffer => {
  if (typeof password !== "string" || password.length === 0) {
    throw new Error("PASSWORD_REQUIRED");
  }
  const encoded = Buffer.from(password, "utf8");
  if (encoded.byteLength > MAX_PASSWORD_BYTES) throw new Error("SCHEMA_INVALID");
  return encoded;
};

export const hashPassword = (
  password: string,
  cost: number = COST_N,
  block: number = BLOCK_R,
  parallel: number = PARALLEL_P,
): string => {
  const salt = randomBytes(SALT_BYTES);
  const key = scryptSync(passwordBytes(password), salt, KEY_BYTES, {
    N: cost,
    r: block,
    p: parallel,
    maxmem: 132 * cost * block,
  });
  return [ALGORITHM, String(cost), String(block), String(parallel), encode(salt), encode(key)].join("$");
};

/**
 * Constant-time verification of a stored digest.
 *
 * A malformed stored value, an unknown algorithm or a non-string input is a
 * failed verification rather than an exception: the caller is an authentication
 * seam that must fail closed.
 */
export const verifyPassword = (password: string, encoded: string): boolean => {
  if (typeof encoded !== "string" || encoded.length === 0) return false;
  const parts = encoded.split("$");
  if (parts.length !== 6 || parts[0] !== ALGORITHM) return false;
  const cost = Number(parts[1]);
  const block = Number(parts[2]);
  const parallel = Number(parts[3]);
  if (!Number.isSafeInteger(cost) || !Number.isSafeInteger(block) || !Number.isSafeInteger(parallel)) return false;
  if (cost < 2 || block < 1 || parallel < 1) return false;
  let salt: Buffer;
  let expected: Buffer;
  try {
    salt = decode(parts[4]!);
    expected = decode(parts[5]!);
  } catch {
    return false;
  }
  if (salt.byteLength === 0 || expected.byteLength !== KEY_BYTES) return false;
  let candidate: Buffer;
  try {
    candidate = scryptSync(passwordBytes(password), salt, KEY_BYTES, {
      N: cost,
      r: block,
      p: parallel,
      maxmem: 132 * cost * block,
    });
  } catch {
    return false;
  }
  return candidate.byteLength === expected.byteLength && timingSafeEqual(candidate, expected);
};
